### DISCUSSION TITLE
Js7 health-checks

### DISCUSSION THREAD

**Discussion by ptr1120** [sos-berlin/js7#15] [Q&A] [answered]
Hello,

are there any liveness and readiness probes available to automatically determine the state of the docker containers?

**Comment by sosap** [comment, marked as answer]
Such checks are available from the [JS7 - REST Web Service API](https://kb.sos-berlin.com/display/JS7/JS7+-+REST+Web+Service+API) and more concsely from the [JS7 - Unix Shell CLI](https://kb.sos-berlin.com/display/JS7/JS7+-+Unix+Shell+Command+Line+Interface)

This implies that checks of Controllers, Agents and JOC Cockpit are performed by requests to the JOC Cockpit container that holds the REST Web Service API. Checks for liveliness/readiness in a JS7 Cluster suggest close reading of the docs.

### Controller

- Scope:
  - Liveness: check _componentState_ returning 0, or 1 severity value
  - Readiness: check _componentState_  and _connectionState_ returning 0 severify value
- Example:
```
response=$(./operate-controller.sh status --url=http://localhost:4446 --user=root --password=root --controller-id=controller)
echo "$response" | jq -r '.controller.componentState.severity // empty'
echo "$response" | jq -r '.controller.connectionState.severity // empty'
```
- Docs: [JS7 - Unix Shell CLI for Controller Status Operations](https://kb.sos-berlin.com/display/JS7/JS7+-+Unix+Shell+CLI+for+Controller+Status+Operations)

### Agent

- Scope:
  - Liveness: check _state_ returning 0, 1 or 2 severity value
  - Readiness: check _state_ returning 0 severify value
- Example:
```
response=$(./operate-controller.sh status-agent --url=http://localhost:4446 --user=root --password=root --controller-id=controller --agent-id=StandaloneAgent)
echo "$response" | jq -r '.agents[0].state.severity // empty'
```
- Docs: [JS7 - Unix Shell CLI for Controller Status Operations](https://kb.sos-berlin.com/display/JS7/JS7+-+Unix+Shell+CLI+for+Controller+Status+Operations)

### JOC Cockpit

- Scope:
  - Liveness: check _componentState_ for the JOC Cockpit instance returning 0 severity value
  - Readiness: check _connectionState_ for the JOC Cockpit returning 0 severity value, check database  _componentState_ and _connectionState_ returning 0 severity value
- Example:
```
response=$(./operate-joc.sh status --url=http://localhost:4446 --user=root --password=root --controller-id=controller)
echo "$response" | jq -r '.jocs[0].componentState.severity // empty'
echo "$response" | jq -r '.jocs[0].connectionState.severity // empty'
echo "$response" | jq -r '.database.componentState.severity // empty'
echo "$response" | jq -r '.database.connectionState.severity // empty'
```
- Docs: [JS7 - Unix Shell CLI for JOC Cockpit Status Operations](https://kb.sos-berlin.com/display/JS7/JS7+-+Unix+Shell+CLI+for+JOC+Cockpit+Status+Operations)

### Simplified Liveness Checks

It is possible to directly access the Controller's and Agent's web service from the container's outside port:

- Controller: `curl 'http://localhost:4444/controller/api'`
- Agent: `curl 'http://localhost:4445/agent/api'`

**Comment by ptr1120** [comment]
Perfect, thank for rhe detailed solution.
